Implemented match history saving feature

Taanviir opened the pull request.

Taanviir committed:
chore: added comments to urls.py

Taanviir committed:
feat: added match.py to views/

Taanviir committed:
feat: updated playerRegister modal to make player4 optional for
tournament

Taanviir committed:
feat: sending game results to backend

Taanviir committed:
feat: updated save_match_result view

MalikZu committed:
fix: get_all_matches of user

MalikZu committed:
feat: update match to have enums for win/loss + override save

MalikZu committed:
feat: add ability to update_stats of user + create match

MalikZu committed:
chore: add Match Model Tests

MalikZu committed:
fix: match model should be name instead of user

MalikZu committed:
chore: update CustomUser to use string for match functions

MalikZu committed:
chore: update tests for match history

Taanviir committed:
feat: match history is being saved now

Taanviir committed:
feat: preventing duplicate usernames in games

MalikZu commented on backend/users/urls.py:
```diff
@@ -2,18 +2,24 @@
     path("2fa/disable", views.disable_2fa, name="disable_2fa"),
     path("2fa/reset", views.reset_2fa_view, name="reset_2fa"),
-    path("login_42/", api_42.login_42, name="login_42"),
+    ## Account Settings Paths
```
removing login_42 route is intentional?

Taanviir committed:
feat: match history implemented for 1v1, 2v2 and tictactoe

- updated profile page to show stats and match history
- handling case where user plays in guest mode (no result is being
  saved)

MalikZu committed:
feat: opponents are array now

MalikZu committed:
chore:update User model

MalikZu committed:
chore: update views function

MalikZu committed:
chore: update tests

MalikZu commented:
now templates should be updated if needed.

Taanviir committed:
chore: ran formatter

Taanviir committed:
Merge branch 'main' into feat/match-history

Taanviir committed:
fix: sending arrays of winners/losers in all cases to backend

- updated profile page to show match history
